### Import libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Import data

In [124]:
path = 'https://raw.githubusercontent.com/dan1dr/microdosing/main/data/dlongcodes.csv'
data = pd.read_csv(path, sep=',', index_col=False)
pd.options.display.max_columns = None

In [171]:
data.head(10)

ResponseId         StartDate           EndDate      RecordedDate  \
0  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
1  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
2  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
3  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
4  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
5  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
6  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
7  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
8  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
9  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   

   MoodBoard_1  MoodBoard_2  MoodBoard_3  MoodBoard_4  MoodBoard_5  \
0            0            1            1            1            1   
1            0            1            1            1            1   
2            0            1            1            1            1   
3            0            1            1            1            1   
4            0            1            1            1            1   
5            0            1            1            1            1   
6            0            0            0            1            0   
7            0            0            0            1            0   
8            0            0            0            1            0   
9            0            0            0            1            0   

   MoodBoard_6  MoodBoard_7  MoodBoard_8  MoodBoard_9  MoodBoard_10  \
0            1            1            1            0             0   
1            1            1            1            0             0   
2            1            1            1            0             0   
3            1            1            1            0             0   
4            1            1            1            0             0   
5            1            1            1            0             0   
6            0            0            0            0             0   
7            0            0            0            0             0   
8            0            0            0            0             0   
9            0            0            0            0             0   

   MoodBoard_11  MoodBoard_12  MoodBoard_13  MoodBoard_14  MoodBoard_15  \
0             1             0             1             0             0   
1             1             0             1             0             0   
2             1             0             1             0             0   
3             1             0             1             0             0   
4             1             0             1             0             0   
5             1             0             1             0             0   
6             0             0             0             0             0   
7             0             0             0             0             0   
8             0             0             0             0             0   
9             0             0             0             0             0   

   MoodBoard_16  MoodBoard_17  MoodBoard_18  MoodBoard_19  MoodBoard_20  \
0             0             1             1             0             1   
1             0             1             1             0             1   
2             0             1             1             0             1   
3             0             1             1             0             1   
4             0             1             1             0             1   
5             0             1             1             0             1   
6             0             1             0             0             0   
7             0             1             0             0             0   
8             0             1             0             0             0   
9             0       

In [122]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1668 entries, 0 to 1667
Columns: 242 entries, ResponseId to ConCategory
dtypes: float64(138), int64(64), object(40)
memory usage: 3.1+ MB


In [123]:
data.shape

(1668, 242)

In [120]:
data.columns

Index(['ResponseId', 'StartDate', 'EndDate', 'RecordedDate', 'MoodBoard_1',
       'MoodBoard_2', 'MoodBoard_3', 'MoodBoard_4', 'MoodBoard_5',
       'MoodBoard_6',
       ...
       'ProRating', 'ResponseId.2', 'ConIndex', 'ConCode', 'Rating.C',
       'ConRating', 'ProConcept', 'ProCategory', 'ConConcept', 'ConCategory'],
      dtype='object', length=242)

## NAN's and duplicates

Since data comes from a survey conducted on the Internet, it may be sometimes be struggling to normalize and clean the format. Thankfully, this dataset was previously used and pre-cleaned by different authors https://osf.io/g5cwy/

However, the format and questionnaire performed is quite complex and we will need to make some adjustments.

Firstly, we need to assess the NANs, duplicates and other elements

In [162]:
def assess_NA(data):
    """
    Returns a pandas dataframe denoting the total number of NA values and the percentage of NA values in each column.
    The column names are noted on the index.
    
    Parameters
    ----------
    data: dataframe
    """
    # pandas series denoting features and the sum of their null values
    null_sum = data.isnull().sum()# instantiate columns for missing data
    total = null_sum.sort_values(ascending=False)
    percent = ( ((null_sum / len(data.index))*100).round(2) ).sort_values(ascending=False)
    
    # concatenate along the columns to create the complete dataframe
    df_NA = pd.concat([total, percent], axis=1, sort=True, keys=['Number of NA', 'Percent NA'])
    
    # drop rows that don't have any missing data; omit if you want to keep all rows
    df_NA = df_NA[ (df_NA.T != 0).any() ]
    
    return df_NA
    
    
    

In [172]:
assess_NA(data)

Number of NA  Percent NA
BFI.2_63                            474       28.42
BFI.2_64                            474       28.42
BFI.2_65                            480       28.78
BFI.2_66                            474       28.42
BFI.2_67                            474       28.42
...                                 ...         ...
RedTobacco                           66        3.96
Reddit_5_TEXT...Topics             1308       78.42
Religion_12_TEXT...Topics          1602       96.04
SES                                  42        2.52
Substance_8_TEXT...Topics          1668      100.00

[160 rows x 2 columns]

Data is relatively "clean" and we will use all the rows given, most of NAs come from fields where the response ask for an alternative text in case of not being listed in the other before, so it is not relevant.

In [126]:
data['ResponseId'].nunique()

278

As we see, the data consists on a set of 278 participants identified by ``ResponseId``. The number of rows is 1668 because the table is given in a "long" format, where the pivot variable is the identity and repeats the same pattern every 6 rows (1668/6 = 278).

This will be important when dealing beyond the 'ProIndex' column, until that point we can dismiss this fact and work only with the unique values in order to acommodate the dataset.

Before splitting the dataset into two different parts, we will rename some columns in order to facilitate future tasks.

In [178]:
data.rename(columns={"MoodBoard_1": "Angry", "MoodBoard_2": "Distressed",
                    "MoodBoard_3": "Afraid", "MoodBoard_4": "Frustrated",
                    "MoodBoard_5": "Nervous", "MoodBoard_6": "Alarmed",
                    "MoodBoard_7": "Tense", "MoodBoard_8": "Annoyed",
                    "MoodBoard_9": "Astonished", "MoodBoard_10": "Excited",
                    "MoodBoard_11": "Alert", "MoodBoard_12": "Delighted",
                    "MoodBoard_13": "Amused", "MoodBoard_14": "Glad",
                    "MoodBoard_15": "Happy", "MoodBoard_16": "Pleased",
                    "MoodBoard_17": "Sad", "MoodBoard_18": "Miserable",
                    "MoodBoard_19": "Depressed", "MoodBoard_20": "Droopy",
                    "MoodBoard_21": "Gloomy", "MoodBoard_22": "Fatigued",
                    "MoodBoard_23": "Bored", "MoodBoard_24": "Tired",
                    "MoodBoard_25": "Serene", "MoodBoard_26": "Content",
                    "MoodBoard_27": "Calm", "MoodBoard_28": "At ease",
                    "MoodBoard_29": "Satisfied", "MoodBoard_30": "Relaxed",
                    "MoodBoard_31": "Sleepy", "MoodBoard_32": "Quiet"}, errors="raise")

ResponseId         StartDate           EndDate      RecordedDate  \
0     R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
1     R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
2     R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
3     R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
4     R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
...                 ...               ...               ...               ...   
1663  R_ZfYg5JxACJX7szf  18/09/2017 14:22  18/09/2017 14:40  18/09/2017 14:40   
1664  R_ZfYg5JxACJX7szf  18/09/2017 14:22  18/09/2017 14:40  18/09/2017 14:40   
1665  R_ZfYg5JxACJX7szf  18/09/2017 14:22  18/09/2017 14:40  18/09/2017 14:40   
1666  R_ZfYg5JxACJX7szf  18/09/2017 14:22  18/09/2017 14:40  18/09/2017 14:40   
1667  R_ZfYg5JxACJX7szf  18/09/2017 14:22  18/09/2017 14:40  18/09/2017 14:40   

      Angry  Distressed  Afraid  Frustrated  Nervous  Alarmed  Tense  Annoyed  \
0         0           1       1           1        1        1      1        1   
1         0           1       1           1        1        1      1        1   
2         0           1       1           1        1        1      1        1   
3         0           1       1           1        1        1      1        1   
4         0           1       1           1        1        1      1        1   
...     ...         ...     ...         ...      ...      ...    ...      ...   
1663      0           0       0           0        0        0      0        0   
1664      0           0       0           0        0        0      0        0   
1665      0           0       0           0        0        0      0        0   
1666      0           0       0           0        0        0      0        0   
1667      0           0       0           0        0        0      0        0   

      Astonished  Excited  Alert  Delighted  Amused  Glad  Happy  Pleased  \
0              0        0      1          0       1     0      0        0   
1              0        0      1          0       1     0      0        0   
2              0        0      1          0       1     0      0        0   
3              0        0      1          0       1     0      0        0   
4              0        0      1          0       1     0      0        0   
...          ...      ...    ...        ...     ...   ...    ...      ...   
1663           0        0      0          0       0     0      0        0   
1664           0        0      0          0       0     0      0        0   
1665           0        0      0          0       0     0      0        0   
1666           0        0      0          0       0     0      0        0   
1667           0        0      0          0       0     0      0        0   

      Sad  Miserable  Depressed  Droopy  Gloomy  Fatigued  Bored  Tired  \
0       1          1          0       1       1         0      0      1   
1       1          1          0       1       1         0      0      1   
2       1          1          0       1       1         0      0      1   
3       1          1          0       1       1         0      0      1   
4       1          1          0       1       1         0      0      1   
...   ...        ...        ...     ...     ...       ...    ...    ...   
1663    0          0          0       0       0         1      0      0   
1664    0          0          0       0       0         1      0      0   
1665    0          0          0       0       0         1      0      0   
1666    0          0          0       0       0         1      0      0   
1667    0          0          0       0       0         1      0      0   

      Serene  Content  Calm  At ease  Satisfied  Relaxed  Sleepy  Quiet  \
0          0        0     0        0          0        0       1      0   
1          0        0     0        0          0        0       1      0   
2          0    

In [127]:
b = data.drop_duplicates(subset = 'ResponseId')
b.shape

(278, 242)

In [17]:
b.index

Int64Index([   0,    6,   12,   18,   24,   30,   36,   42,   48,   54,
            ...
            1608, 1614, 1620, 1626, 1632, 1638, 1644, 1650, 1656, 1662],
           dtype='int64', length=278)

In [149]:
#Rearrange the index from 0
b.index = range(1,279)
b.head()

ResponseId         StartDate           EndDate      RecordedDate  \
1  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
2  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
3  R_0N8EpmmQ5b79cxr  18/10/2017 20:29  18/10/2017 23:51  18/10/2017 23:51   
4  R_0Nj0uZDZlPP0IRX  18/09/2017 11:40  18/09/2017 11:58  18/09/2017 11:58   
5  R_0OZWDFj2krlpms9  22/09/2017 18:03  23/09/2017 06:47  07/10/2017 06:47   

   MoodBoard_1  MoodBoard_2  MoodBoard_3  MoodBoard_4  MoodBoard_5  \
1            0            1            1            1            1   
2            0            0            0            1            0   
3            0            0            0            0            0   
4            0            0            0            0            0   
5            0            1            0            0            1   

   MoodBoard_6  MoodBoard_7  MoodBoard_8  MoodBoard_9  MoodBoard_10  \
1            1            1            1            0             0   
2            0            0            0            0             0   
3            0            0            0            0             0   
4            0            0            0            0             0   
5            0            0            0            0             1   

   MoodBoard_11  MoodBoard_12  MoodBoard_13  MoodBoard_14  MoodBoard_15  \
1             1             0             1             0             0   
2             0             0             0             0             0   
3             0             0             0             0             1   
4             0             0             0             0             0   
5             1             0             1             0             1   

   MoodBoard_16  MoodBoard_17  MoodBoard_18  MoodBoard_19  MoodBoard_20  \
1             0             1             1             0             1   
2             0             1             0             0             0   
3             0             0             0             0             1   
4             0             0             0             0             1   
5             1             0             0             0             1   

   MoodBoard_21  MoodBoard_22  MoodBoard_23  MoodBoard_24  MoodBoard_25  \
1             1             0             0             1             0   
2             0             1             1             1             0   
3             0             1             0             1             0   
4             0             1             0             0             0   
5             0             0             0             1             1   

   MoodBoard_26  MoodBoard_27  MoodBoard_28  MoodBoard_29  MoodBoard_30  \
1             0             0             0             0             0   
2             1             1             0             0             1   
3             1             0             0             0             1   
4             0             0             0             0             0   
5             1             0             1             1             1   

   MoodBoard_31  MoodBoard_32  Current  Reddit  Country  Age  Gender  \
1             1             0        1       5      187   21    -1.0   
2             0             1        1       1      187   29    -1.0   
3             1             0        2       2      187   20    -1.0   
4             0             0        1       1      128   31    -1.0   
5             1             0        2       5       -1   27    -1.0   

   Orientation  Ethnicity Ethnicity_11_TEXT  Religion Religion_12_TEXT  \
1         -1.0          1               -99         1              -99   
2         -1.0          6               -99         1              -99   
3         -1.0         13               -99         1              -99   
4         -1.0          1               -99         1              -99   
5         -1.0          1               -99        14              -99   

   Religios

Beyond the 'ProIndex' columns, we find the reason of the long format. Exploring the original data https://osf.io/dwtk5/ we find that there are represented the following values: the first three correspond to "pro_text_X_topics", and the last three to "pro_X_code".

The first 3 rows are almost identical to the following 3, just differ when sometimes one is NA or the topics is different from the category. We will select the category as it is a more general characteristic. Hence, we will only use the last 3 rows per each ``ResponseId``

In [168]:
long_info = data.loc[:, 'ProIndex':]
long_info.head(12)

ProIndex               ProCode       ResponseId.1  Rating.P  ProRating  \
0          1         Introspection  R_0HSeYpblz5LCLpD         1       85.0   
1          2  Perspective Shifting  R_0HSeYpblz5LCLpD         2       90.0   
2          3                   NaN  R_0HSeYpblz5LCLpD         3       55.0   
3          4         Introspection  R_0HSeYpblz5LCLpD         4       85.0   
4          5  Perspective Shifting  R_0HSeYpblz5LCLpD         5       90.0   
5          6             Self-Care  R_0HSeYpblz5LCLpD         6       55.0   
6          1            Creativity  R_0MpphsGkrwSMhuF         1       70.0   
7          2   Improved Motivation  R_0MpphsGkrwSMhuF         2       52.0   
8          3         Improved Mood  R_0MpphsGkrwSMhuF         3       45.0   
9          4            Creativity  R_0MpphsGkrwSMhuF         4       70.0   
10         5   Improved Motivation  R_0MpphsGkrwSMhuF         5       52.0   
11         6         Improved Mood  R_0MpphsGkrwSMhuF         6       45.0   

         ResponseId.2  ConIndex             ConCode  Rating.C  ConRating  \
0   R_0HSeYpblz5LCLpD         1  Sleep Disturbances         1       25.0   
1   R_0HSeYpblz5LCLpD         2          Illegality         2        4.0   
2   R_0HSeYpblz5LCLpD         3              Stigma         3       20.0   
3   R_0HSeYpblz5LCLpD         4  Sleep Disturbances         4       25.0   
4   R_0HSeYpblz5LCLpD         5          Illegality         5        4.0   
5   R_0HSeYpblz5LCLpD         6              Stigma         6       20.0   
6   R_0MpphsGkrwSMhuF         1  Sleep Disturbances         1       35.0   
7   R_0MpphsGkrwSMhuF         2                 NaN         2        NaN   
8   R_0MpphsGkrwSMhuF         3                 NaN         3        NaN   
9   R_0MpphsGkrwSMhuF         4  Sleep Disturbances         4       35.0   
10  R_0MpphsGkrwSMhuF         5                 NaN         5        NaN   
11  R_0MpphsGkrwSMhuF         6                 NaN         6        NaN   

                 ProConcept    ProCategory                  ConConcept  \
0                 Self-Care  Self-Efficacy  Physiological Interference   
1   Meta-Creative Processes     Creativity                  Illegality   
2                       NaN            NaN                      Stigma   
3                 Self-Care  Self-Efficacy  Physiological Interference   
4   Meta-Creative Processes     Creativity                  Illegality   
5                 Self-Care  Self-Efficacy                      Stigma   
6                Creativity     Creativity  Physiological Interference   
7             Self-Efficacy  Self-Efficacy                         NaN   
8             Improved Mood  Improved Mood                         NaN   
9                Creativity     Creativity  Physiological Interference   
10            Self-Efficacy  Self-Efficacy                         NaN   
11            Improved Mood  Improved Mood                         NaN   

                 ConCategory  
0   Physiological Discomfort  
1                 Illegality  
2                 Illegality  
3   Physiological Discomfort  
4                 Illegality  
5                 Illegality  
6   Physiological Discomfort  
7                        NaN  
8                        NaN  
9   Physiological Discomfort  
10                       NaN  
11                       NaN

The strategy to conduct the data analysis is the following:

- Using the shortdata for exploring most characteristics
- Using the longdata when assessing the general Pros and Cons per each user

# Data Wrangling